### ISSUE TITLE
Atomic memory order not keep consistent?

### ISSUE THREAD

**Issue by zhongwuzw** [bignerdranch/Deferred#194]
`bnr_atomic_ptr_compare_and_swap` uses `__ATOMIC_ACQ_REL` memory order , but all other `load` commands use `__ATOMIC_RELAXED` that not have synchronization each other, so `load` maybe read `nil` but actually the value is not `nil` .  

i.e.
`group.notify` block called because of `bnr_atomic_ptr_compare_and_swap` success, but `load` operation in `group.notify` block maybe read `nil`, leaded to `body(Storage.unbox(from: ptr))` not called.

I think `load` operation should use `__ATOMIC_ACQUIRE` to build synchronization relation .

**Comment by zwaldowski** [comment]
None of the `load` calls are intended to have a specific ordering relative to one another. They are only ever called in response to the unblocking of the dispatch group following a successful `fill` or a `deinit`.

As an aside, the standard indicates an `ACQ` should be followed by a `REL`, so it is not appropriate for independent reads.

**Comment by zhongwuzw** [comment]
@zwaldowski , `isFilled`、`wait` are all public API , we can call one in any thread. 

> the standard indicates an ACQ should be followed by a REL

Sorry, I don't understand what's your mean, would you explain more? In my opinion, `ACQ` and `REL` are pairwise. when `bnr_atomic_ptr_compare_and_swap` success, it generate a `REL` command, the meantime `load` can immediately read latest value because of `REL-ACQ` `synchronizes-with` in concurrently environment(`load` must be called after `write` but almost meantime). 

[《Concurrency in action》 chapter 5.2.2](http://www.bogotobogo.com/cplusplus/files/CplusplusConcurrencyInAction_PracticalMultithreading.pdf)
[Cpp reference](http://en.cppreference.com/w/cpp/atomic/atomic_flag)

**Comment by zwaldowski** [comment]
> `isFilled`, `wait` are all public API , we can call one in any thread.

`wait` uses the group to wait, so its call to `load` is blocked on another thread’s successful `fill` as I describe above.

`isFilled` is only meant to be a “best effort”, not a guarantee. It shouldn’t be used defensively before attempting to `fill`. There is no thread ordering that will prevent a `fill` from happening on another thread in between a call to `isFilled` and a call to `fill` on the same thread. The most accurate way to check will always be attempting to `fill`.

**Comment by zwaldowski** [comment]
Using `ACQUIRE` for `isFilled` does sound legitimate, though, and I can look into it further.

**Comment by zhongwuzw** [comment]
@zwaldowski  `bnr_atomic_ptr_compare_and_swap` uses `__ATOMIC_ACQ_REL` when success, it's not necessary because all other atomic operations besides `load` are `__ATOMIC_RELAXED`, there are not memory order issue. So `bnr_atomic_ptr_compare_and_swap` can always use `__ATOMIC_RELAXED` .

**Comment by zwaldowski** [comment]
By my reading of cppreference and the book you link to, I don’t think either of those conclusions are right. We are only interested in defining a strong ordering relationship when competing for the one-time initialization of storage, which by every resource I’m finding is still an appropriate use of `ACQ_REL`. Some of the orderings of loads can be debated, but as I establish above we don’t seek strong guarantees for those.

I suspect this discussion will lead us in circles. To pick up this topic again, I ask for a test case displaying a behavior you’re not satisfied with. In the meantime, I don’t want to pose any risks to our production users by making changes to this area.

Thank you so much for the discussion!

**Comment by zhongwuzw** [comment]
:beers::beers::beers:
My last comment, Haha. 
the reason of the resource you find use `ACQ_REL` is that they use `ACQ` for `load` in other thread. `__ATOMIC_RELAXED` don't participate in synchronizes-with relationships .
![image](https://user-images.githubusercontent.com/5061845/30147015-49a24a88-93ce-11e7-915b-c3286922b90d.png)
